distributed/worker.py

```diff
@@ -920,3 +717,666 @@ def run(worker, stream, function=None, args=(), kwargs={}):
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb; pdb.set_trace()
+            raise
+
+    def transition_executing_done(self, key):
+        try:
+            if self.validate:
+                assert key in self.executing or key in self.long_running
+                assert key not in self.waiting_for_data
+                assert key not in self.heap
+
+            if self.task_state[key] == 'executing':
+                self.executing.remove(key)
+            elif self.task_state[key] == 'long-running':
+                self.long_running.remove(key)
+            if self.batched_stream:
+                self.batched_stream.send(self.response[key])
+
+        except StreamClosedError:
+            logger.info("Stream closed")
+            self._close(report=False)
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb; pdb.set_trace()
+            raise
+
+    def transition_executing_long_running(self, key):
+        try:
+            if self.validate:
+                assert key in self.executing
+
+            self.executing.remove(key)
+            self.long_running.add(key)
+
+            self.ensure_computing()
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb; pdb.set_trace()
+            raise
+
+    ##########################
+    # Gather Data from Peers #
+    ##########################
+
+    def ensure_communicating(self):
+        try:
+            while self.data_needed and len(self.connections) < self.total_connections:
+                logger.debug("Ensure communicating.  Pending: %d.  Connections: %d/%d",
+                             len(self.data_needed),
+                             len(self.connections),
+                             self.total_connections)
+
+                key = self.data_needed[0]
+
+                if key not in self.tasks:
+                    self.data_needed.popleft()
+                    continue
+
+                if self.task_state.get(key) != 'waiting':
+                    self.log.append((key, 'communication pass'))
+                    self.data_needed.popleft()
+                    continue
+
+                deps = self.dependencies[key]
+                deps = [d for d in deps
+                          if d not in self.data
+                          and d not in self.executing
+                          and d not in self.in_flight]
+
+                for dep in deps:
+                    if not self.who_has.get(dep):
+                        logger.info("Can't find dependencies for key %s", key)
+                        self.cancel_key(key)
+                        continue
+
+                self.log.append(('gather-dependencies', key, deps))
+
+                while deps and len(self.connections) < self.total_connections:
+                    self.gather_dep(deps.pop())
```

Comment:
I plan to switch this out though for something else, just to improve our ability to profile the worker.  Ideally we can isolate most of the logic into non-tornado code.